File: packages/core/src/bacchant.ts
import { BASE64_VALUES, BACCHANT_SYMBOL_THRESHOLDS, BLOCK_SIZE, HEADER_SIZE } from './constants';

export interface DecodedResult {
  payload: string | null;
  status: string;
}

export class BacchantCodec {
  /**
   * Decodes a Bacchant payload from ImageData.
   */
  public decode(imageData: ImageData): DecodedResult {
    const { width, height, data } = imageData;

    const getBin = (x: number, y: number): number => {
      let total = 0;
      for (let i = 0; i < BLOCK_SIZE; i++) {
        for (let j = 0; j < BLOCK_SIZE; j++) {
          const idx = ((y + j) * width + (x + i)) * 4;
          if (idx + 2 >= data.length) return 0;
          const r = data[idx];
          const g = data[idx + 1];
          const b = data[idx + 2];
          const lum = 0.299 * r + 0.587 * g + 0.114 * b;
          total += lum;
        }
      }
      const avg = total / (BLOCK_SIZE * BLOCK_SIZE);
      const inc = 255.0 / 7.0;
      const bin = Math.round(avg / inc);
      return Math.min(7, Math.max(0, bin));
    };

    // 1. Verify Header
    let header = "";
    for (let y = 0; y < HEADER_SIZE; y += BLOCK_SIZE) {
      for (let x = 0; x < HEADER_SIZE; x += 2 * BLOCK_SIZE) {
        const b0 = getBin(x, y);
        const b1 = getBin(x + BLOCK_SIZE, y);
        header += BASE64_VALUES.charAt(b0 * 8 + b1);
      }
    }

    if (header !== "bacchant") {
      return { payload: null, status: "invalid header" };
    }

    // 2. Decode Length and Payload
    let lengthChars = "";
    let payloadChars = "";
    let payloadLen: number | null = null;
    let charCount = 0;

    for (let y = 0; y < height; y += BLOCK_SIZE) {
      for (let x = 0; x < width; x += 2 * BLOCK_SIZE) {
        if (y < HEADER_SIZE && x < HEADER_SIZE) {
          continue;
        }

        const b0 = getBin(x, y);
        const b1 = getBin(x + BLOCK_SIZE, y);
        const ch = BASE64_VALUES.charAt(b0 * 8 + b1);

        if (charCount < 8) {
          lengthChars += ch;
          charCount++;
        } else {
          if (payloadLen === null) {
            payloadLen = parseInt(lengthChars, 10);
            if (isNaN(payloadLen)) {
              return { payload: null, status: "invalid length" };
            }
          }
          payloadChars += ch;
          if (payloadChars.length >= payloadLen) {
            return { payload: payloadChars.substring(0, payloadLen), status: "ok" };
          }
        }
      }
    }

    return { payload: payloadChars, status: "incomplete" };
  }

  /**
   * Encodes a payload into ImageData (simulated via Uint8ClampedArray).
   */
  public encode(payload: string, width: number, height: number): Uint8ClampedArray {
    const data = new Uint8ClampedArray(width * height * 4);
    data.fill(128);

    const setBlock = (x: number, y: number, symbol: number) => {
      const level = BACCHANT_SYMBOL_THRESHOLDS[symbol];
      for (let i = 0; i < BLOCK_SIZE; i++) {
        for (let j = 0; j < BLOCK_SIZE; j++) {
          const idx = ((y + j) * width + (x + i)) * 4;
          if (idx + 3 < data.length) {
            data[idx] = level;
            data[idx + 1] = level;
            data[idx + 2] = level;
            data[idx + 3] = 255;
          }
        }
      }
    };

    const getOctals = (char: string): [number, number] => {
      const idx = BASE64_VALUES.indexOf(char);
      if (idx === -1) return [0, 0]; // Fallback for non-base64
      return [Math.floor(idx / 8), idx % 8];
    };

    // 1. Encode Header
    const headerStr = "bacchant";
    for (let i = 0; i < headerStr.length; i++) {
      const [s0, s1] = getOctals(headerStr[i]);
      const sym0Idx = 2 * i;
      const sym1Idx = 2 * i + 1;
      const x0 = (sym0Idx % 4) * BLOCK_SIZE;
      const y0 = Math.floor(sym0Idx / 4) * BLOCK_SIZE;
      setBlock(x0, y0, s0);
      const x1 = (sym1Idx % 4) * BLOCK_SIZE;
      const y1 = Math.floor(sym1Idx / 4) * BLOCK_SIZE;
      setBlock(x1, y1, s1);
    }

    // 2. Encode Payload
    const lengthStr = payload.length.toString().padStart(8, '0');
    const fullData = lengthStr + payload;
    const symbols: number[] = [];
    for (let i = 0; i < fullData.length; i++) {
      const [s0, s1] = getOctals(fullData[i]);
      symbols.push(s0, s1);
    }

    const nHeaderRowSymbolsWide = (width - HEADER_SIZE) / BLOCK_SIZE;
    const nSymbolsInFullRow = width / BLOCK_SIZE;

    symbols.forEach((sym, i) => {
      let x, y;
      if (i < nHeaderRowSymbolsWide * 4) {
        const yCoord = Math.floor(i / nHeaderRowSymbolsWide);
        const xCoord = i % nHeaderRowSymbolsWide;
        x = HEADER_SIZE + (xCoord * BLOCK_SIZE);
        y = yCoord * BLOCK_SIZE;
      } else {
        const i2 = i + 16;
        const yCoord = Math.floor(i2 / nSymbolsInFullRow);
        const xCoord = i2 % nSymbolsInFullRow;
        x = xCoord * BLOCK_SIZE;
        y = yCoord * BLOCK_SIZE;
      }

      if (x + BLOCK_SIZE <= width && y + BLOCK_SIZE <= height) {
        setBlock(x, y, sym);
      }
    });

    return data;
  }
}

File: core-rust/src/bacchant.rs
use crate::constants::*;

pub struct BacchantCodec;

impl BacchantCodec {
    pub fn decode(width: usize, height: usize, data: &[u8]) -> (Option<String>, String) {
        let get_bin = |x: usize, y: usize| -> u8 {
            let mut total: f64 = 0.0;
            for i in 0..BLOCK_SIZE {
                for j in 0..BLOCK_SIZE {
                    let idx = ((y + j) * width + (x + i)) * 4;
                    if idx + 2 < data.len() {
                        let r = data[idx] as f64;
                        let g = data[idx + 1] as f64;
                        let b = data[idx + 2] as f64;
                        let lum = 0.299 * r + 0.587 * g + 0.114 * b;
                        total += lum;
                    }
                }
            }
            let avg = total / (BLOCK_SIZE * BLOCK_SIZE) as f64;
            let inc = 255.0 / 7.0;
            let bin = (avg / inc).round() as u8;
            bin.min(7)
        };

        // 1. Verify Header
        let mut header = String::new();
        for y in (0..HEADER_SIZE).step_by(BLOCK_SIZE) {
            for x in (0..HEADER_SIZE).step_by(2 * BLOCK_SIZE) {
                let b0 = get_bin(x, y);
                let b1 = get_bin(x + BLOCK_SIZE, y);
                let idx = (b0 * 8 + b1) as usize;
                if let Some(ch) = BASE64_VALUES.chars().nth(idx) {
                    header.push(ch);
                }
            }
        }

        if header != "bacchant" {
            return (None, "invalid header".to_string());
        }

        // 2. Decode Length and Payload
        let mut length_chars = String::new();
        let mut payload_chars = String::new();
        let mut payload_len: Option<usize> = None;
        let mut count = 0;

        for y in (0..height).step_by(BLOCK_SIZE) {
            for x in (0..width).step_by(2 * BLOCK_SIZE) {
                if y < HEADER_SIZE && x < HEADER_SIZE {
                    continue;
                }

                let b0 = get_bin(x, y);
                let b1 = get_bin(x + BLOCK_SIZE, y);
                let idx = (b0 * 8 + b1) as usize;
                let ch = BASE64_VALUES.chars().nth(idx).unwrap_or('?');

                if y == 0 && x < HEADER_SIZE + 16 * BLOCK_SIZE {
                    length_chars.push(ch);
                } else {
                    if payload_len.is_none() {
                        payload_len = length_chars.parse::<usize>().ok();
                        if payload_len.is_none() {
                            return (None, "invalid length".to_string());
                        }
                    }
                    
                    if let Some(len) = payload_len {
                        payload_chars.push(ch);
                        count += 1;
                        if count >= len {
                            return (Some(payload_chars), "ok".to_string());
                        }
                    }
                }
            }
        }

        (Some(payload_chars), "incomplete".to_string())
    }

    pub fn encode(payload: &str, width: usize, height: usize) -> Vec<u8> {
        let mut data = vec![128u8; width * height * 4];
        for i in 0..width * height {
            data[i * 4 + 3] = 255; // Opaque
        }

        let mut set_block = |x: usize, y: usize, symbol: u8| {
            let level = BACCHANT_SYMBOL_THRESHOLDS[symbol as usize];
            for i in 0..BLOCK_SIZE {
                for j in 0..BLOCK_SIZE {
                    let idx = ((y + j) * width + (x + i)) * 4;
                    data[idx] = level;
                    data[idx + 1] = level;
                    data[idx + 2] = level;
                }
            }
        };

        let get_octals = |ch: char| -> (u8, u8) {
            let idx = BASE64_VALUES.find(ch).unwrap_or(0);
            ((idx / 8) as u8, (idx % 8) as u8)
        };

        // 1. Header
        let header_str = "bacchant";
        for (i, ch) in header_str.chars().enumerate() {
            let (s0, s1) = get_octals(ch);
            let x = (i % 4) * BLOCK_SIZE;
            let y = (i / 4) * BLOCK_SIZE;
            set_block(x, y, s0);
            set_block(x + BLOCK_SIZE, y, s1);
        }

        // 2. Payload
        let length_str = format!("{:08}", payload.len());
        let full_data = format!("{}{}", length_str, payload);
        let mut symbols = Vec::new();
        for ch in full_data.chars() {
            let (s0, s1) = get_octals(ch);
            symbols.push(s0);
            symbols.push(s1);
        }

        let n_header_row_symbols_wide = (width - HEADER_SIZE) / BLOCK_SIZE;
        let n_symbols_in_full_row = width / BLOCK_SIZE;

        for (i, &sym) in symbols.iter().enumerate() {
            let (x, y);
            if i < n_header_row_symbols_wide * 4 {
                let y_coord = i / n_header_row_symbols_wide;
                let x_coord = i % n_header_row_symbols_wide;
                x = HEADER_SIZE + (x_coord * BLOCK_SIZE);
                y = y_coord * BLOCK_SIZE;
            } else {
                let i2 = i + 16;
                let y_coord = i2 / n_symbols_in_full_row;
                let x_coord = i2 % n_symbols_in_full_row;
                x = x_coord * BLOCK_SIZE;
                y = y_coord * BLOCK_SIZE;
            }

            if x + BLOCK_SIZE <= width && y + BLOCK_SIZE <= height {
                set_block(x, y, sym);
            }
        }

        data
    }
}
